File: ASD/es_2/max_level.c
#include <stdlib.h>
#include <stdio.h>

#include "queue.h"
#include "max_level.h"
#include "tree.h"

int max(int a, int b) { return a > b ? a : b; }

int max_level(node_tree * n) {
    node_tree* s = n;
    int keys_sum = get_node_info(s);
    while ((s = get_next_sibling(s)) != NULL)
        keys_sum += get_node_info(s);

    node_tree* first_child = get_first_child(n);
    if (first_child == NULL) return keys_sum;
    return max(keys_sum, max_level(first_child));
}

File: ASD/Exams/20150914/problem_2/java-sol/MaxLevel.java
public class MaxLevel {

    public static int maxLevel(NodeTree n) {

        if (n == null)
            return -1;

        Queue q = new Queue();
        Queue next_q = new Queue();
        NodeTree node = n;

        int depth = 0;
        int max_level = 0;
        int max_sum = 0;

        int sum = 0;

        while(node != null) {

            sum += node.getNodeInfo();

            if (node.getFirstChild() != null)
                next_q.enqueue(node.getFirstChild());

            if (node.getNextSibling() != null)
                node = node.getNextSibling();
            else
                node = null;

            if (node == null && q.size() > 0)
                node = q.dequeue();

            if (node == null) {

                System.out.println("Visited all nodes at level " + depth + " - sum: " + sum);
                if (sum > max_sum) {
                    max_level = depth;
                    max_sum = sum;
                }

                depth += 1;

                sum = 0;

                Queue temp = q;
                q = next_q;
                next_q = temp;

                node = q.dequeue();
            }
        }

        return max_level;
    }

}
